crates/sui-adapter/src/execution_engine.rs

```diff
@@ -504,6 +529,25 @@ fn advance_epoch<S: BackingPackageStore + ParentSync + ChildObjectResolver>(
         )
         .expect("Advance epoch with safe mode must succeed");
     }
+    // MUSTFIX: The code below needs to be gated by a feature flag for a new protocol version.
+    {
+        let unmetered_storage_rebate = gas_status.unmetered_storage_rebate();
+        debug!(
+            "Amount of unmetered storage rebate from advance epoch tx: {:?}",
+            unmetered_storage_rebate
+        );
+        // Put all the storage rebate accumulated in the system transaction
+        // to the 0x5 object so that it's not lost.
+        let mut system_state_wrapper = temporary_store
+            .read_object(&SUI_SYSTEM_STATE_OBJECT_ID)
+            .expect("0x5 object must be muated in advance_epoch tx")
+            .clone();
+        // In unmetered execution, storage_rebate field of mutated object must be 0.
+        // If not, we would be dropping SUI on the floor by overriding it.
+        assert_eq!(system_state_wrapper.storage_rebate, 0);
```

Comment:
What is the intuition for why this will be 0?

Reply:
This is the mutated state, not the original state. And when we mutate an object, we reset its storage_rebate (indicating that a user is paying for the new storage). But since in this case nobody is paying, its new storage_rebate must be 0

Reply:
This makes sense to me as a high-level explanation. 

But in the current gas charging logic, I'm not sure there's a place where we reset the object's `storage_rebate` to 0 before setting it to its new value--I think we directly [set](https://github.com/MystenLabs/sui/blob/main/crates/sui-types/src/temporary_store.rs#L838) it to the new value. I'm also not sure how unmetered code interacts with all of this. I think @dariorussi will know better.